spec/function.dd

```diff
@@ -2015,6 +2015,48 @@ S test6(ref return scope int* p)
 }
 ---
 
+$(H4 $(LNAME2 pure-scope-inference, Inferred `scope` parameters in `pure` functions))
+
+    When a parameter is not marked or inferred `scope`, it may still be `@safe` to assign it a `scope` pointer in a function call.
+    The following conditions need to be met:
+
+    $(UL
+    $(LI The function is $(RELATIVE_LINK2 pure-functions, `pure`), hence the argument cannot be assigned to a global variable)
+    $(LI The function is $(RELATIVE_LINK2 nothrow-functions, `nothrow`), hence the argument cannot be assigned to a thrown `Exception` object)
+    $(LI None of the other parameters have mutable indirections, hence the argument cannot be assigned to a longer-lived variable)
```

Comment:
Just a query: what about that feature where if the function returns void then `return` applies to the first parameter? I don't know if that's still supported (if so could be added separately).

Reply:
`return` semantics when assigning to the first parameter is still supported, but `pure` -> `scope` inference doesn't consider it yet: it sees the first parameter is mutable, and won't infer (return) scope for the other parameters. This is a possible enhancement to `pure` -> `scope` inference.